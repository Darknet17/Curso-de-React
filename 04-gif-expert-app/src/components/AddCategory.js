import {useState} from "react";
import PropTypes from 'prop-types';
import {Button, Form, FormControl} from "react-bootstrap";

export const AddCategory = ({setCategories}) => {

    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        if(inputValue.trim().length > 2){
            setCategories(cat=>[inputValue, ...cat])
            setInputValue('');
        }

    }

    return (
        <Form className="d-flex" onSubmit={ handleSubmit }>
            <FormControl
                type="search"
                placeholder="Search"
                className="me-2"
                aria-label="Search"
                value={ inputValue }
                onChange={ handleInputChange }
            />
            <Button type="submit" variant="outline-success">Search</Button>
        </Form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
}

