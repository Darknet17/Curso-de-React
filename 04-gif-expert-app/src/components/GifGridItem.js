import PropTypes from "prop-types";
import {Card} from "react-bootstrap";

export const GifGridItem = ({ title, url }) => {
    return (
        <Card className="card animate__animated animate__fadeIn" style={{ width: '18rem' }}>
            <Card.Img variant="top" src={ url } />
            <Card.Body>
                <Card.Text> {
                    title
                }</Card.Text>
            </Card.Body>
        </Card>
    );
}

GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url:  PropTypes.string.isRequired
}
