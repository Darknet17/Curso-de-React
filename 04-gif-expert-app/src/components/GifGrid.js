import PropTypes from "prop-types";
import {GifGridItem} from "./GifGridItem";
import {useFetchGifs} from "../hooks/useFetchGifs";
import {Col, Row, Spinner} from "react-bootstrap";

export const GifGrid = ({ category }) => {

    const {data: images, loading} = useFetchGifs(category);

    return (
        <>
            <h3 className="animate__animated animate__fadeIn">{category}</h3>

            {  loading && <Spinner animation="grow" variant="dark" />  }

            <Row>
                {
                    images.map(  ({id, title, url }) => (
                        <Col sm={3}>
                            <GifGridItem key={id} title={title} url={url}/>
                        </Col>
                    ))
                }
            </Row>
        </>
    );
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}


