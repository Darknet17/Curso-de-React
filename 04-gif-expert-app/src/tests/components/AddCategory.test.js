import { shallow } from "enzyme";
import {AddCategory} from "../../components/AddCategory";

describe("Probando el <AddCategory />", ()=>{

    const setCategories = jest.fn();
    let wrapper;

    beforeEach(()=>{
       jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories={setCategories}/>);
    });

    test("debe de mostrar el componente",()=>{
        expect(wrapper).toMatchSnapshot();
    });

    test("debe de cambiar la caja de texto",()=>{
        const input = wrapper.find('FormControl');
        const value = "Hola Mundo";
        input.simulate('change', { target: { value } });
        // expect(wrapper.find("p").text().trim()).toBe(value);
    });

    test("No debe de postear la informacion con el submit",()=>{
        wrapper.find("Form").simulate('submit', { preventDefault: () => {} });
        expect(setCategories).not.toHaveBeenCalled();
    });

    test("debe de llamar el setCategories y limpiar la caja de texto", ()=>{
        // 1. Simular el inputChange
        // 2. Simular el submit
        // 3. setCategories se debe de haber llamado
        // 4. El valor del input debe de estar ''

        const value = "Hola Mundo";
        const input = wrapper.find("FormControl");
        // 1. Simular el inputChange
        input.simulate('change', { target: { value } });

        // 2. Simular el submit setCategories se debe de haber llamado
        wrapper.find("Form").simulate('submit', { preventDefault: () => {} });

        // 3. setCategories se debe de haber llamado
        expect(setCategories).toHaveBeenCalled();
        expect(setCategories).toHaveBeenCalledTimes(1);
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function));

        // 4. El valor del input debe de estar ''
        expect( wrapper.find("FormControl").prop("value") ).toBe("");
    });

});
