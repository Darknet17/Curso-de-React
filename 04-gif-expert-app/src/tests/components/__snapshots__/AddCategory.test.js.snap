// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando el <AddCategory /> debe de mostrar el componente 1`] = `
<Form
  className="d-flex"
  onSubmit={[Function]}
>
  <FormControl
    aria-label="Search"
    className="me-2"
    onChange={[Function]}
    placeholder="Search"
    type="search"
    value=""
  />
  <Button
    active={false}
    disabled={false}
    type="submit"
    variant="outline-success"
  >
    Search
  </Button>
</Form>
`;
